cache/contenthash/checksum_test.go

```diff
@@ -434,6 +434,114 @@ func TestChecksumBasicFile(t *testing.T) {
 	require.NoError(t, err)
 }
 
+func TestChecksumIncludeExclude(t *testing.T) {
+	t.Parallel()
+	tmpdir, err := ioutil.TempDir("", "buildkit-state")
+	require.NoError(t, err)
+	defer os.RemoveAll(tmpdir)
+
+	snapshotter, err := native.NewSnapshotter(filepath.Join(tmpdir, "snapshots"))
+	require.NoError(t, err)
+	cm, _ := setupCacheManager(t, tmpdir, "native", snapshotter)
+	defer cm.Close()
+
+	ch := []string{
+		"ADD foo file data0",
+		"ADD bar file data1",
+		"ADD d0 dir",
+		"ADD d0/abc file abc",
+		"ADD d1 dir",
+		"ADD d1/def file def",
+	}
+
+	ref := createRef(t, cm, ch)
+
+	cc, err := newCacheContext(ref.Metadata(), nil)
+	require.NoError(t, err)
+
+	dgst, err := cc.Checksum(context.TODO(), ref, "foo", ChecksumOpts{IncludePatterns: []string{"foo"}}, nil)
```

Comment:
What do you think the behavior should be in this case? Should we exclude `/foo` because the specified path is not a dir? Return an error?

Reply:
Same as copying a directory where `IncludePatterns` are set but none match.

Reply:
If I understand correctly, you think it should copy no files in this scenario?

I don't think that matches the behavior implemented in `fsutil`. `(*copier).copy` only checks `IncludePatterns` on recursive calls, so calling `Copy` with `/foo` would copy `/foo` regardless of `IncludePatterns`. We would need a PR to fsutils to keep the behavior in sync.

Reply:
So dir with no matching subdirs will make an empty dir? Thanks seems ok.

Not sure about the file case. What do you think is the best? It is always wrong to use it like this and a sign that the user did something wrong. Do you see cases where it might seem wrong for the user if they get an error for this case?

Reply:
It seems reasonable to me to make it an error to try and do a `Copy` op with `IncludePatterns` or `ExcludePatterns` when the source is a file. If we do that, we don't need any special handling at the checksum level, since it wouldn't be a valid scenario.

I could add a `Stat` in `docopy` if you think that makes sense. Or we could add this check at the `fsutil` level.